pkg/query/table.go

```diff
@@ -0,0 +1,392 @@
+	"github.com/parca-dev/parca/pkg/profile"
+)
+
+const (
+	TableFieldMappingStart   = "mapping_start"
+	TableFieldMappingLimit   = "mapping_limit"
+	TableFieldMappingOffset  = "mapping_offset"
+	TableFieldMappingFile    = "mapping_file"
+	TableFieldMappingBuildID = "mapping_build_id"
+
+	TableFieldLocationAddress = "location_address"
+	TableFieldLocationFolded  = "location_folded"
+	TableFieldLocationLine    = "location_line"
+
+	TableFieldFunctionStartLine  = "function_startline"
+	TableFieldFunctionName       = "function_name"
+	TableFieldFunctionSystemName = "function_system_name"
+	TableFieldFunctionFileName   = "function_file_name"
+
+	TableFieldCumulative     = "cumulative"
+	TableFieldCumulativeDiff = "cumulative_diff"
+	TableFieldFlat           = "flat"
+	TableFieldFlatDiff       = "flat_diff"
+)
+
+func GenerateTable(
+	ctx context.Context,
+	mem memory.Allocator,
+	tracer trace.Tracer,
+	p profile.Profile,
+) (*queryv1alpha1.TableArrow, int64, error) {
+	ctx, span := tracer.Start(ctx, "GenerateTable")
+	defer span.End()
+
+	record, cumulative, err := generateTableArrowRecord(ctx, mem, tracer, p)
+	if err != nil {
+		return nil, 0, err
+	}
+	defer record.Release()
+
+	// TODO: Reuse buffer and potentially writers
+	var buf bytes.Buffer
+	w := ipc.NewWriter(&buf,
+		ipc.WithSchema(record.Schema()),
+		ipc.WithAllocator(mem),
+	)
+	defer w.Close()
+
+	if err = w.Write(record); err != nil {
+		return nil, 0, err
+	}
+
+	return &queryv1alpha1.TableArrow{
+		Record: buf.Bytes(),
+		Unit:   p.Meta.SampleType.Unit,
+	}, cumulative, nil
+}
+
+func generateTableArrowRecord(
+	ctx context.Context,
+	mem memory.Allocator,
+	tracer trace.Tracer,
+	p profile.Profile,
+) (arrow.Record, int64, error) {
+	_, span := tracer.Start(ctx, "generateTableArrowRecord")
+	defer span.End()
+
+	tb := newTableBuilder(mem)
+	defer tb.Release()
+
+	row := 0
+
+	profileReader := profile.NewReader(p)
+	for _, r := range profileReader.RecordReaders {
+		tb.cumulative += math.Int64.Sum(r.Value)
+
+		for sampleRow := 0; sampleRow < int(r.Record.NumRows()); sampleRow++ {
+			lOffsetStart, lOffsetEnd := r.Locations.ValueOffsets(sampleRow)
+			for locationRow := int(lOffsetStart); locationRow < int(lOffsetEnd); locationRow++ {
+				if r.Lines.IsNull(locationRow) {
+					// The location has no lines, we therefore compare its address.
+
+					isLeaf := locationRow == int(lOffsetStart)
+					var buildID []byte
+					if r.MappingBuildIDDict.IsValid(locationRow) {
+						buildID = r.MappingBuildIDDict.Value(r.MappingBuildID.GetValueIndex(locationRow))
+					}
+					addr := r.Address.Value(locationRow)
+
+					// Check if we've seen the address for the mapping before.
+					// If not, we add it as a new row and add the address to the mapping to keep track of it.
+					// If we have seen the address before, we merge the address with the existing row by summing the values.
+					if cr, ok := tb.addresses[unsafeString(buildID)][addr]; !ok {
```

Comment:
where do we ensure that this doesn't panic?

Reply:
Investigating this a bit. Seems like this works out fine. If the first check isn't successful the second one doesn't get checked.
```go
func TestNestedMapPanic(t *testing.T) {
    foo := map[int]map[int]struct{}{}

    if _, ok := foo[0][0]; ok {
        fmt.Println("found")
    } else {
        fmt.Println("not found")
    }
}
```